audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -38,6 +46,8 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
 
+        //get run-time permission...
+        requestResourcePermission();
```

Comment:
Recommended UX here would be to request in context, ie. when clicking the button.
I agree that this will potentially complicate the flow of the app, but how about this:
- Request in onClickListener of record button:
    - If permission is missing, request it
    - If permission has been granted, continue as usual
- In onRequestPermissionsResult callback:
   - If granted, continue as usual 
   - If denied, display a dialog explaining that permissions are required, with a button to re-request permission or close the app

I'd be happy for you to move the permission request into onCreate, that way you don't need to handle different flows through the app (and handle the state where you are missing permissions. You could then remove the permissionGranted flag.)

We have done exactly this for some of our other platform samples too. See the relevant code here: https://github.com/googlesamples/android-Camera2Basic/commit/241c6fac8113ddb5dffb4f349b5bb9f1bea93501